### PR TITLE
h2port config changes 

### PR DESCRIPTION
**Pull request by jamesbr3**
This introduces a new configuration option for h2 database access. The aim is to have stronger security by default, however current behaviour can be restored easily:

* Maintains the existing h2Port config option. However the behaviour is slighty different and the node listens on localhost instead of 0.0.0.0 (which is more secure) 
* The new syntax is h2Settings { address : "host:port" }
* If h2Port and h2Settings are both specified, the node will throw an error at startup
* If neither is specified, the node *will not* expose its H2 database. This is a more significant change because currently the node will always expose its H2 database even if this is not requested in the config.

Documention for node-administration and node-database has been updated
Changelog updated.

### CODE REVIEW

**Comment by mnesbit** [comment]
Do we wish to sync up cordformation?

File: docs/source/node-administration.rst [hunk 1]
```diff
@@ -72,6 +71,19 @@ Any database browsing tool that supports JDBC can be used, but if you have Intel
 a tool integrated with your IDE. Just open the database window and add an H2 data source with the above details.
 You will now be able to browse the tables and row data within them.
 
+By default the node will expose its database the localhost network interface. This behaviour can be
```
Comment (tudor-malene) [inline docs/source/node-administration.rst]: ..its database "to" the localhost network interface

File: docs/source/node-administration.rst [hunk 2]
```diff
@@ -72,6 +71,19 @@ Any database browsing tool that supports JDBC can be used, but if you have Intel
...
+overridden by using specifying the full network address (interface and port) using the new h2Settings
```
Comment (tudor-malene) [inline docs/source/node-administration.rst]: -using

File: docs/source/node-administration.rst [hunk 3]
```diff
@@ -72,6 +71,19 @@ Any database browsing tool that supports JDBC can be used, but if you have Intel
...
+syntax in the node configuration:
+
+.. sourcecode:: groovy
+  h2Settings {
+      address: "localhost:12345"
+  }
+
+The configuration above will restrict the H2 service to run on localhost. If remote access is required, the address
+can be changed to 0.0.0.0. However it is recommended to change the default username and password
```
Comment (tudor-malene) [inline docs/source/node-administration.rst]: Do we support changing the default h2 user/pass?

File: docs/source/node-database.rst [hunk 4]
```diff
@@ -25,6 +32,9 @@ stored states, transactions and attachments as follows:
 You will be presented with a web interface that shows the contents of your node's storage and vault, and provides an
 interface for you to query them using SQL.
 
+The default behaviour is to expose the H2 database on the localhost interface. This can be overridden in the
+node configuration us
```
Comment (tudor-malene) [inline docs/source/node-database.rst]: This seems unfinished

File: node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt [hunk 5]
```diff
@@ -277,6 +297,10 @@ data class NodeRpcSettings(
     }
 }
 
+data class NodeH2Settings(
+        val address: NetworkHostAndPort?
+) {}
```
Comment (tudor-malene) [inline node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt]: no need for curly braces

File: tools/demobench/src/test/kotlin/net/corda/demobench/model/NodeConfigTest.kt [hunk 6]
```diff
@@ -45,7 +45,7 @@ class NodeConfigTest {
         assertEquals(localPort(40002), fullConfig.rpcOptions.address)
         assertEquals(localPort(10001), fullConfig.p2pAddress)
         assertEquals(listOf(user("jenny")), fullConfig.rpcUsers)
-        assertThat(fullConfig.dataSourceProperties["dataSource.url"] as String).contains("AUTO_SERVER_PORT=30001")
+//        assertThat(fullConfig.dataSourceProperties["dataSource.url"] as String).contains("AUTO_SERVER_PORT=30001")
```
Comment (tudor-malene) [inline tools/demobench/src/test/kotlin/net/corda/demobench/model/NodeConfigTest.kt]: Should be removed if no longer needed

**Comment by jamesbr3** [comment]
@mnesbit Yes that is the intention (to sync up with CordFormation). This change is intended to be backward compatible;  CordFormation will continue to work because it specifies an explicit h2Port parameter. 

I was going to submit a subsequent change to for CordFormation and NodeDriver so that those components can start using the newer h2 config block.

File: docs/source/node-administration.rst (hunk 3 above)
Comment (jamesbr3) [reply docs/source/node-administration.rst]: AFAIK the user/pass is specified on the h2 database, not through the TCP server that we expose. So we specify that using existing dataSourceProperties config block

**Comment by jamesbr3** [comment]
@mnesbit @tudor-malene note that I have introduced a further change to the default H2 datasource URL. (in reference.conf) so that the ALLOW_LITERALS setting is set to NONE. This is to make sql injection more difficult.

File: node/src/main/resources/reference.conf [hunk 7]
```diff
@@ -5,15 +5,15 @@ crlCheckSoftFail = true
 lazyBridgeStart = true
 dataSourceProperties = {
     dataSourceClassName = org.h2.jdbcx.JdbcDataSource
-    dataSource.url = "jdbc:h2:file:"${baseDirectory}"/persistence;DB_CLOSE_ON_EXIT=FALSE;WRITE_DELAY=0;LOCK_TIMEOUT=10000;AUTO_SERVER_PORT="${h2port}
+    dataSource.url = "jdbc:h2:file:"${baseDirectory}"/persistence;DB_CLOSE_ON_EXIT=FALSE;LOCK_TIMEOUT=10000;WRITE_DELAY=100"
```
Comment (tudor-malene) [inline node/src/main/resources/reference.conf]: WRITE_DELAY=100 needs to be replaced with WRITE_DELAY=0.
This setting caused lots of grief to @exFalso

Comment (jamesbr3) [reply node/src/main/resources/reference.conf]: ok fixed, sorry must have merged with a stale config

**Pull request merged**